cloudevents/extensions/opcua.md

```diff
@@ -0,0 +1,335 @@
+Header](https://reference.opcfoundation.org/Core/Part14/v105/docs/7.2.5.4#Table164)
+field `DataSetWriterId` or `DataSetWriterName`.
+
+For event messages (type equals to `ua-event`) `subject` MUST additional contain 
+"/" and [Base Event Type](https://reference.opcfoundation.org/Core/Part5/v104/docs/6.4.2)
+field `EventId`. 
+
+### time
+
+MUST map to [Data Set Message
+Header](https://reference.opcfoundation.org/Core/Part14/v105/docs/7.2.5.4#Table164)
+field `Timestamp`.
+
+## Mapping for other extensions
+
+The following well-known extensions attributes are used for data messages and
+event messages (type one of `ua-keyframe`, `ua-deltaframe`, `ua-event`).
+
+### sequence
+
+Attribute as defined by [sequence extensions](./sequence.md) MUST map to [Data
+Set Message Header](https://reference.opcfoundation.org/Core/Part14/v105/docs/7.2.5.4#Table164)
+field `SequenceNumber`.
+
+### traceparent
+
+Attribute as defined by [distributed-tracing extension](./distributed-tracing.md)
+MUST be used to allow tracing from event publisher towards consumer.
+
+### tracestate
+
+Attribute as defined by [distributed-tracing extension](./distributed-tracing.md)
+MAY be used to allow tracing from event publisher towards consumer.
+
+### recordedtime
+
+Attribute as defined by [recordedtime extension](./recordedtime.md) MUST be used
+to determine the latency between event publisher towards consumer. 
+
+## Attributes
+
+### opcuametadatamajorversion
+
+- Type: `Integer`
+- Description: Links dataset message to the current version of the metadata.
+Contains value from `MajorVersion` of [Data Set Message Header](https://reference.opcfoundation.org/Core/Part14/v105/docs/7.2.5.4#Table164) field `MetaDataVersion`. 
+- Constraints
+  - OPTIONAL but MUST NOT be present if `dataschema` is used
+
+### opcuametadataminorversion
+
+- Type: `Integer`
+- Description: Links dataset message to the current version of the metadata.
+Contains value from `MiniorVersion` of [Data Set Message
+Header](https://reference.opcfoundation.org/Core/Part14/v105/docs/7.2.5.4#Table164)
+field `MetaDataVersion`.
+- Constraints
+  - OPTIONAL but MUST NOT be present if `dataschema` is used
+
+### opcuastatus
+
+- Type: `Integer`
+- Description: Defines the overall status of the data set message, maps to 
+[Data Set Message Header](https://reference.opcfoundation.org/Core/Part14/v105/docs/7.2.5.4#Table164) field `Status`.
+- Constraints
+  - OPTIONAL
+  - MAY be omitted if status is _Good_
+
+## General Constraints
+
+- OPC UA messages MUST use `binary-mode` of CloudEvents.
+- OPC UA PubSub JSON messages MUST be encoded using non-reversible encoding as
+the decoding information is contained in metadata messages or by schema
+referenced via `dataschema` attribute.
+- Payload of OPC UA PubSub JSON messages MUST NOT contain Network Message Header
+and Data Set Header as those information are mapped into CloudEvents attributes.
+- OPC UA PubSub JSON messages MUST only contain one dataset message.
```

Comment:
can they contain zero? If yes then perhaps "MUST not contain more than one....". If no, then perhaps "MUST contain exactly one..."

Reply:
In OPC UA an Network Message can contain multiple dataset messages, as we "promote" dataset and network header fields to CloudEvents attributes, there can only be exactly one contained. Changed to "MUST contain exactly one"